docs: expand mixins information, notate body reuse (#1209)
nodejs/undici: README.md

console.log('trailers', trailers)
```

## Body Mixins

> The `body` mixins are the most common way to format the request/response body from a [`ReadableStream`](https://streams.spec.whatwg.org/#readablestream). Mixins include:

Reviewer: I don't think it's correct. the request method does not have anything to do with WHATWG `ReadableStream`.

Author: totally open to suggestion. was a best-stab at it without intimate knowledge, but to show how we could be thorough

Reviewer: I would just remove the reference to ReadableStream.

Author: ```suggestion
The `body` mixins are the most common way to format the request/response body. Mixins include:
```

Author: @ronag please reopen this so I can apply the suggestion above

console.log('trailers', trailers)
```

> _Note: Once a mixin has been called, the `ReadableStream` the body represents cannot be reused, thus calling additional mixins on `.body`, e.g. `.body.json(); .body.text()` will result in an error `TypeError: unusable` being thrown and returned through the `Promise` rejection._
> 

Reviewer 2: ```suggestion
_Note: Once a mixin has been called, the the body cannot be reused, thus calling additional mixins on `.body`, e.g. `.body.json(); .body.text()` will result in an error `TypeError: unusable` being thrown and returned through the `Promise` rejection._

```

Author: You all are welcome to finish this out the way you see best, or close this one. No offense taken either way. I feel that I've been super accommodating and the process has just gotten a bit too pedantic for my liking. 🍻

console.log('trailers', trailers)
```

_Note: Once a mixin has been called, the `ReadableStream` the body represents cannot be reused, thus calling additional mixins on `.body`, e.g. `.body.json(); .body.text()` will result in an error `TypeError: unusable` being thrown and returned through the `Promise` rejection._

Should you need to access the `body` in plain-text after using a mixin, the best practice is to use the `.text()` mixin first, and manually parse the text to the desired format.

> If the scenario arises where you absolutely must have a copy of the result stream, [`.clone`](https://fetch.spec.whatwg.org/#dom-request-clone) can be used _before_ using a `body` mixin, or [`body.pipeThrough`](https://streams.spec.whatwg.org/#readablestream-pipe-through) with a user-defined stream [§](https://web.dev/fetch-upload-streaming), such as [`TransformStream`]https://nodejs.org/api/stream.html#duplex-and-transform-streams.

Reviewer: cc @ronag how do you think this can be fixed?

Reviewer 2: What do you mean by fixed?

Reviewer: I mean that this refer to `body.pipeThrough()` about the `http.request()` while we do not have a WHATWG stream there.

Reviewer 2: I don't understand what the problem is.

Reviewer 2: Like you suggested above. Mentions to web streams should be removed IMHO.

Author: That would really neuter the usefulness of the paragraph. I'd really like to retain something for users to point them in the right direction. Could you suggest an edit here?

Reviewer 2: ReadableStream or anything web stream related is not relevant here. I think maybe you want to put another version of this under fetch?

Author: OK, I can remove it. I don't think it belong under `fetch` because users that need to copy the result stream _while using undici_ (and that's the important part here) should have something to reference. I understand that from a maintainer's point of view it doesn't belong, but from a user's point of view, that takes away a resource of understanding.

If we can't canonize it in the docs, would it be acceptable to open a new issue as a "how to,"  paste the contents above in the issue, and then close it? So at the least there'd be something searchable in the repo.

Reviewer 2: >  but from a user's point of view, that takes away a resource of understanding.

I'm sorry bit I'm not sure how it belongs here even from a user's point of view. I find it misleading since there is no web stream in this part of the API.

> If we can't canonize it in the docs, would it be acceptable to open a new issue as a "how to," paste the contents above in the issue, and then close it? So at the least there'd be something searchable in the repo.

Sure...

> If the scenario arises where you absolutely must have a copy of the result stream, [`.clone`](https://fetch.spec.whatwg.org/#dom-request-clone) can be used _before_ using a `body` mixin, or [`body.pipeThrough`](https://streams.spec.whatwg.org/#readablestream-pipe-through) with a user-defined stream [§](https://web.dev/fetch-upload-streaming), such as [`TransformS tream`]https://nodejs.org/api/stream.html#duplex-and-transform-streams.

Reviewer 2: ```suggestion
If the scenario arises where you absolutely must have a copy of the result stream, [`.clone`](https://fetch.spec.whatwg.org/#dom-request-clone) can be used _before_ using a `body` mixin, or [`body.pipeThrough`](https://streams.spec.whatwg.org/#readablestream-pipe-through) with a user-defined stream [§](https://web.dev/fetch-upload-streaming), such as [`TransformStream`]https://nodejs.org/api/stream.html#duplex-and-transform-streams.
```
